### PR TITLE
Fix issues found by static analysis

### PR DESCRIPTION
**Pull request by JelteF**

### CODE REVIEW

File: src/bin/pg_autoctl/monitor.c [hunk 1]
```diff
@@ -661,6 +661,12 @@ printCurrentState(void *ctx, PGresult *result)
 	/* prepare a nice dynamic string of '-' as a header separator */
 	nameSeparatorHeader = (char *) malloc((maxNodeNameSize+1) * sizeof(char));
 
+	if (nameSeparatorHeader == NULL) {
+		log_error("could not allocate");
+		context->parsedOK = false;
+		return;
+	}
+
```
Comment (DimCitus) [inline src/bin/pg_autoctl/monitor.c]: Style issue: brackets on their own lines.

File: src/bin/pg_autoctl/pgctl.c [hunk 2]
```diff
@@ -181,6 +181,10 @@ config_find_pg_ctl(PostgresSetup *pgSetup)
 	{
 		char *program = pg_ctls[0];
 		char *version = pg_ctl_version(program);
+		if (version == NULL) {
+			log_fatal("Found %s, but failed to get version", program);
+			return 0;
+		}
```
Comment (DimCitus) [inline src/bin/pg_autoctl/pgctl.c]: Style issue, brackets on their own line.

File: src/bin/pg_autoctl/monitor.c [hunk 3]
```diff
@@ -661,6 +661,12 @@ printCurrentState(void *ctx, PGresult *result)
 	/* prepare a nice dynamic string of '-' as a header separator */
 	nameSeparatorHeader = (char *) malloc((maxNodeNameSize+1) * sizeof(char));
 
+	if (nameSeparatorHeader == NULL) {
+		log_error("could not allocate");
```
Comment (DimCitus) [inline src/bin/pg_autoctl/monitor.c]: Improve error message to something the user might understand and know what do to about...

File: src/bin/pg_autoctl/pgctl.c [hunk 4]
```diff
@@ -181,6 +181,10 @@ config_find_pg_ctl(PostgresSetup *pgSetup)
 	{
 		char *program = pg_ctls[0];
 		char *version = pg_ctl_version(program);
+		if (version == NULL) {
+			log_fatal("Found %s, but failed to get version", program);
```
Comment (DimCitus) [inline src/bin/pg_autoctl/pgctl.c]: I think I would rephrase as "Failed to get output from `%s version`". It's a very rare event, but at least showing what we tried might allow the user to copy/paste and see for themselves, either the same problem, or that we have a bug somewhere.

File: src/bin/pg_autoctl/pgctl.c [hunk 5]
```diff
@@ -199,6 +203,10 @@ config_find_pg_ctl(PostgresSetup *pgSetup)
 		{
 			char *program = pg_ctls[i];
 			char *version = pg_ctl_version(program);
+			if (version == NULL) {
+				log_warn("Found %s, but failed to get version", program);
+				continue;
+			}
```
Comment (DimCitus) [inline src/bin/pg_autoctl/pgctl.c]: Style issue: brackets on their own lines.

File: src/bin/pg_autoctl/pgctl.c [hunk 6]
```diff
@@ -199,6 +203,10 @@ config_find_pg_ctl(PostgresSetup *pgSetup)
 		{
 			char *program = pg_ctls[i];
 			char *version = pg_ctl_version(program);
+			if (version == NULL) {
+				log_warn("Found %s, but failed to get version", program);
```
Comment (DimCitus) [inline src/bin/pg_autoctl/pgctl.c]: Same error message comment as before.

File: src/monitor/group_state_machine.c [hunk 7]
```diff
@@ -62,6 +62,9 @@ ProceedGroupState(AutoFailoverNode *activeNode)
 {
 	AutoFailoverNode *otherNode = NULL;
 	AutoFailoverFormation *formation = GetFormation(activeNode->formationId);
+	if (formation == NULL) {
+		ereport(ERROR, (errmsg("Formation for %s could not be found", activeNode->formationId)));
+	}
```
Comment (DimCitus) [inline src/monitor/group_state_machine.c]: Style issue: brackets on their own line.

File: src/monitor/node_active_protocol.c [hunk 8]
```diff
@@ -170,6 +170,14 @@ register_node(PG_FUNCTION_ARGS)
...
 
 	pgAutoFailoverNode = GetAutoFailoverNode(nodeName, nodePort);
+	if (pgAutoFailoverNode == NULL) {
+		ereport(ERROR,
+				(errcode(ERRCODE_INTERNAL_ERROR),
+				 errmsg("node %s:%d with dbname \"%s\" could not be registered in "
+						"formation \"%s\", could not get matching node",
+						nodeName, nodePort, expectedDBName,
+						formationId)));
+	}
```
Comment (DimCitus) [inline src/monitor/node_active_protocol.c]: Style issue: brackets on their own lines.

File: src/monitor/node_active_protocol.c [hunk 9]
```diff
@@ -170,6 +170,14 @@ register_node(PG_FUNCTION_ARGS)
 	LockNodeGroup(formationId, currentNodeState.groupId, ExclusiveLock);
 
 	pgAutoFailoverNode = GetAutoFailoverNode(nodeName, nodePort);
+	if (pgAutoFailoverNode == NULL) {
+		ereport(ERROR,
+				(errcode(ERRCODE_INTERNAL_ERROR),
+				 errmsg("node %s:%d with dbname \"%s\" could not be registered in "
+						"formation \"%s\", could not get matching node",
```
Comment (DimCitus) [inline src/monitor/node_active_protocol.c]: Let's improve the error message. The node we're failing to find now is the one we just did an INSERT for in the previous lines of the code...

File: src/bin/pg_autoctl/file_utils.c [hunk 10]
```diff
@@ -119,6 +119,24 @@ ensure_empty_dir(const char *dirname, int mode)
...
+ * sure files are not accidentally created with umask 777 if the user has it
+ * configured in a weird way.
+ */
+FILE *
+fopen_with_umask(const char *filePath, const char* modes, int flags, mode_t umask)
+{
+	int fileDescriptor = open(filePath, O_WRONLY | O_CREAT, umask);
+	if (fileDescriptor == -1)
+	{
+		return NULL;
```
Comment (DimCitus) [inline src/bin/pg_autoctl/file_utils.c]: A comment saying that we don't need to log anything would be good...

File: src/bin/pg_autoctl/parsing.c [hunk 11]
```diff
@@ -91,6 +91,11 @@ regexp_first_match(const char *string, const char *regex)
 		int finish = m[1].rm_eo;
 		int length = finish - start + 1;
 		char *result = (char *) malloc(length * sizeof(char));
+		if (result == NULL)
+		{
+			log_error("Failed to allocate memory, probably because it's all full");
```
Comment (DimCitus) [inline src/bin/pg_autoctl/parsing.c]: full? used?

File: src/bin/pg_autoctl/file_utils.c (hunk 10 above)
Comment (JelteF) [reply src/bin/pg_autoctl/file_utils.c]: I modified the function to do the logging instead. I also changed it to clase the fileDescriptor in case `open` succeeds, but fdopen fails.

File: src/bin/pg_autoctl/file_utils.c [hunk 12]
```diff
@@ -119,6 +119,35 @@ ensure_empty_dir(const char *dirname, int mode)
...
+fopen_with_umask(const char *filePath, const char* modes, int flags, mode_t umask)
+{
+	int fileDescriptor = open(filePath, O_WRONLY | O_CREAT, umask);
+	FILE *fileStream = NULL;
+	if (fileDescriptor != -1)
+	{
+		fileStream = fdopen(fileDescriptor, modes);
+		if (fileStream == NULL) {
+			close(fileDescriptor);
+		}
```
Comment (DimCitus) [inline src/bin/pg_autoctl/file_utils.c]: Style issue: brackets on their own lines.

File: src/bin/pg_autoctl/file_utils.c [hunk 13]
```diff
@@ -119,6 +119,35 @@ ensure_empty_dir(const char *dirname, int mode)
...
+	}
+	if (fileStream == NULL) {
+		log_error("Failed to open file \"%s\": %s", filePath, strerror(errno));
+	}
```
Comment (DimCitus) [inline src/bin/pg_autoctl/file_utils.c]: Style issue: brackets on their own lines. Also, more to the point, do we need to check if fileStream is NULL in two different places? We could log_error and return from the first place, I guess?

File: src/bin/pg_autoctl/file_utils.c [hunk 14]
```diff
@@ -127,12 +156,10 @@ ensure_empty_dir(const char *dirname, int mode)
 bool
 write_file(char *data, long fileSize, const char *filePath)
 {
-	FILE *fileStream = NULL;
 
-	fileStream = fopen(filePath, "wb");
+	FILE *fileStream = fopen_with_umask(filePath, "wb", O_WRONLY | O_CREAT, 0644);
 	if (fileStream == NULL)
 	{
-		log_error("Failed to open file \"%s\": %s", filePath, strerror(errno));
```
Comment (DimCitus) [inline src/bin/pg_autoctl/file_utils.c]: Please add a comment about not logging on purpose, as we do everywhere in the code:

``` c
    /* errors have already been logged */
```

File: src/bin/pg_autoctl/file_utils.c [hunk 15]
```diff
@@ -161,12 +188,9 @@ write_file(char *data, long fileSize, const char *filePath)
 bool
 append_to_file(char *data, long fileSize, const char *filePath)
 {
-	FILE *fileStream = NULL;
-
-	fileStream = fopen(filePath, "ab");
+	FILE *fileStream = fopen_with_umask(filePath, "ab", O_APPEND | O_CREAT, 0644);
 	if (fileStream == NULL)
 	{
-		log_error("Failed to open file \"%s\": %s", filePath, strerror(errno));
```
Comment (DimCitus) [inline src/bin/pg_autoctl/file_utils.c]: Please add a comment about not logging on purpose, as we do everywhere in the code:

``` c
    /* errors have already been logged */
```

File: src/bin/pg_autoctl/keeper_config.c [hunk 16]
```diff
@@ -415,10 +415,9 @@ keeper_config_write_file(KeeperConfig *config)
 
 	log_trace("keeper_config_write_file \"%s\"", filePath);
 
-	fileStream = fopen(filePath, "w");
+	fileStream = fopen_with_umask(filePath, "w", O_WRONLY | O_CREAT, 0644);
 	if (fileStream == NULL)
 	{
-		log_error("Failed to open file \"%s\": %s", filePath, strerror(errno));
```
Comment (DimCitus) [inline src/bin/pg_autoctl/keeper_config.c]: Please add a comment about not logging on purpose, as we do everywhere in the code:

``` c
    /* errors have already been logged */
```

File: src/bin/pg_autoctl/monitor_config.c [hunk 17]
```diff
@@ -314,10 +314,9 @@ monitor_config_write_file(MonitorConfig *config)
 
 	log_trace("monitor_config_write_file \"%s\"", filePath);
 
-	fileStream = fopen(filePath, "w");
+	fileStream = fopen_with_umask(filePath, "w", O_WRONLY | O_CREAT, 0644);
 	if (fileStream == NULL)
 	{
-		log_error("Failed to open file \"%s\": %s", filePath, strerror(errno));
```
Comment (DimCitus) [inline src/bin/pg_autoctl/monitor_config.c]: Please add a comment about not logging on purpose, as we do everywhere in the code:

``` c
    /* errors have already been logged */
```

File: src/bin/pg_autoctl/pgctl.c [hunk 18]
```diff
@@ -199,6 +204,11 @@ config_find_pg_ctl(PostgresSetup *pgSetup)
 		{
 			char *program = pg_ctls[i];
 			char *version = pg_ctl_version(program);
+			if (version == NULL)
+			{
+				log_warn("Failed to get version info from %s --version", program);
+				continue;
```
Comment (DimCitus) [inline src/bin/pg_autoctl/pgctl.c]: I think this `continue` here warrants a comment explaining what's happening. We could have been skipping those entries from the PATH search result, for instance, and decided that if there's only one entry left that's the one we're going to use. It looks logical to do that, but I think it would be the wrong thing and your implementation is the right one. A comment would avoid rediscovering the reasoning behind it every time we look at that code.
